[libc++] Add MSVC's implementation of `exception_ptr` for Windows

StephanTLavavej opened the pull request:
As requested by @petrhosek (and earlier by @llvm-beanz), Microsoft is happy to contribute our implementation of `exception_ptr` for Windows, under the Apache License v2.0 with LLVM Exception that our STLs share.

I've copied our files to plausible-looking places and updated their line endings and banners, but otherwise have made no attempt to integrate them into libc++. I copied all of our `<exception>` even though you only need `exception_ptr` from it. These are our current sources, corresponding to what's going to ship for production use in VS 2022 17.11.

Petr explained to me that libc++ developers will be able to do the integration work - please feel free to push arbitrary changes to my branch. I expect that significant overhauls will be necessary! (Significant improvements should be possible when you don't have to worry about the ancient evil `/clr:pure` etc.)

Please let me know if you need any more information or code. There's a bit of `shared_ptr` integration, where that machinery grants access to `_Exception_ptr_access` defined in `excptptr.cpp` - if you need the `<memory>` side of things, I can add that. (For the record, libc++ is always welcome to grab sources directly from microsoft/STL under our license, but we're also happy to deliver sources like this.)

The only caveat is that our "VCRuntime" support library is not open-source. If you need sources from VCRuntime, we'll have to figure out whether that's possible. Presumably this shouldn't be necessary, since our open-source microsoft/STL consumes the proprietary VCRuntime "at arm's length" and everything works.

Finally, although I've maintained our STL since 2007 and have worked on almost every part of it, I didn't write `exception_ptr` and its inner workings remain mysterious to me. It was written by a Microsoft dev who left the VC Libraries team in 2013 (and later, the company). If you have questions about what it's doing, I can try to get answers, but sometimes your guess will be as good as ours. :joy_cat: There's a potential performance improvement that could be made, which we're currently unable to do for ABI compatibility, see https://github.com/microsoft/STL/issues/751. I don't see any active bugs in our internal database about `exception_ptr`.

## Commits

* microsoft/STL's `excptptr.cpp`, copied verbatim.
  + From our `main` branch as of 2024-06-10: https://github.com/microsoft/STL/blob/e36ee6c2b9bc6f5b1f70776c18cf5d3a93a69798/stl/src/excptptr.cpp
* microsoft/STL's `<exception>`, copied verbatim as `win32_exception_ptr.h`.
  + From our `main` branch as of 2024-06-10: https://github.com/microsoft/STL/blob/e36ee6c2b9bc6f5b1f70776c18cf5d3a93a69798/stl/inc/exception
* Convert files to LF line endings.
* Add LLVM banners.
* Update `libcxx/CREDITS.TXT` to mention Microsoft.
  + I am intentionally not crediting myself here, I'm just delivering the pizza, I didn't cook it. :pizza:

StephanTLavavej committed:
microsoft/STL's `excptptr.cpp`, copied verbatim.

From our `main` branch as of 2024-06-10:

https://github.com/microsoft/STL/blob/e36ee6c2b9bc6f5b1f70776c18cf5d3a93a69798/stl/src/excptptr.cpp

StephanTLavavej committed:
Update `libcxx/CREDITS.TXT` to mention Microsoft.

I am intentionally not crediting myself here,
I'm just delivering the pizza, I didn't cook it.

StephanTLavavej committed:
Convert files to LF line endings.

StephanTLavavej committed:
Add LLVM banners.

StephanTLavavej committed:
microsoft/STL's `<exception>`, copied verbatim as `win32_exception_ptr.h`.

From our `main` branch as of 2024-06-10:

https://github.com/microsoft/STL/blob/e36ee6c2b9bc6f5b1f70776c18cf5d3a93a69798/stl/inc/exception

frederick-vs-ja commented on libcxx/include/__exception/win32_exception_ptr.h:
```diff
@@ -0,0 +1,430 @@
+//===----------------------------------------------------------------------===//
```
I think the new header and source files should be added to `libcxx/include/CMakeLists.txt` and `libcxx/src/CMakeLists.txt`.

frederick-vs-ja commented on libcxx/include/__exception/win32_exception_ptr.h:
```diff
@@ -0,0 +1,430 @@
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#include <cstdlib>
+#include <type_traits>
```
Should we include granularized internal headers?

frederick-vs-ja commented on libcxx/include/__exception/win32_exception_ptr.h:
```diff
@@ -0,0 +1,430 @@
+
+#if _HAS_EXCEPTIONS
+
+#include <malloc.h> // TRANSITION, VSO-2048380: This is unnecessary, but many projects assume it as of 2024-04-29
```
Is this also needed for libc++ even though the internal header is newly added?

frederick-vs-ja commented on libcxx/src/support/win32/excptptr.cpp:
```diff
@@ -0,0 +1,541 @@
+        }
+
+    public:
+        // constexpr, TRANSITION P1064
+        explicit _ExceptionPtr_static() noexcept : _Ref_count_base() {
+            _PopulateCppExceptionRecord(_ExRecord, &_Ex, static_cast<ThrowInfo*>(__GetExceptionInfo(_Ex)));
+        }
```
It seems that we can never make this ctor `constexpr` as `_PopulateCppExceptionRecord` needs to perform `reinterpret_cast` and `RtlPcToFileHeader` (even if `__GetExceptionInfo` gets `constexpr`-ized one day).
Should we drop this comment (also in MSVC STL)?

petrhosek committed:
Integrate Microsoft STL exception_ptr implementation into libc++

* Deleted libcxx/include/__exception/win32_exception_ptr.h (which was a
  verbatim copy of Microsoft STL's <exception> header). All standard
  exception classes (exception, bad_alloc, bad_exception,
  nested_exception, etc.) are already natively provided by libc++'s
  standard headers.
* Replaced MSVC STL naming conventions and macros (_STD, _CRTIMP2_PURE,
  _Ref_count_base, _Ptr_base, _EXCEPTION_RECORD, etc.) with standard
  libc++ style and identifiers (_LIBCPP_EXPORTED_FROM_ABI,
  std::__shared_weak_count, __double_leading_underscore
  variable/function names).
* Rewrote reference count control blocks (__exception_ptr_static and
  __exception_ptr_normal) to inherit from std::__shared_weak_count.
* Replaced internal MSVC CRT headers (<trnsctrl.h>, <internal_shared.h>,
  <xcall_once.h>) with portable standard C++ constructs (e.g., standard
  C++ function-local static initialization for immortal static exception
  objects) and inline pointer-to-member invocation helpers.
* Guaranteed 16-byte heap allocation alignment for
  __exception_ptr_normal using standard
  alignas(__STDCPP_DEFAULT_NEW_ALIGNMENT__).

petrhosek committed:
Merge the implementation into exception_pointer_msvc.ipp

* Removed all ten exported __ExceptionPtr... C functions. These symbols
  were originally present in Microsoft's CRT DLL exports solely to
  support inline methods in MSVC STL's <exception> header. User code
  compiled against libc++ public headers calls standard C++ functions
  (current_exception(), rethrow_exception, swap, operator==,
  constructors/destructors) and never references the __ExceptionPtr... C
  runtime symbols.
* Moved all SEH/EH exception record definitions, SEH copying helpers
  (__populate_cpp_exception_record, __copy_exception_object), and
  reference-counted control blocks (__exception_ptr_static and
  __exception_ptr_normal) directly into exception_pointer_msvc.ipp.
* Implemented std::exception_ptr constructors, destructor, assignment
  operators, and queries directly in namespace std via placement new
  (::new (this) shared_ptr<const _EXCEPTION_RECORD>(...)) and
  reinterpret_cast conversion helpers (__to_shared), taking full
  advantage of the guaranteed layout compatibility between exception_ptr
  and shared_ptr<const _EXCEPTION_RECORD>.

petrhosek committed:
Use reference-counted control block representation

* Removed the platform #ifdef _LIBCPP_ABI_MICROSOFT bifurcation for
  class exception_ptr in exception_ptr.h. On all ABI targets (Itanium,
  MSVC, etc.), sizeof(std::exception_ptr) == sizeof(void*) (8 bytes on
  x64).
* std::exception_ptr now shares the exact same single-pointer data
  member (void* __ptr_;), trivial relocatability attribute (using
  __trivially_relocatable), and inline operations (default constructor,
  nullptr constructor, move constructor, move assignment, operator
  bool(), operator==, operator!=, and swap) across all operating
  systems.
* __exception_ptr_storage encapsulates the _EXCEPTION_RECORD __record_;
  member immediately following the std::__shared_count base layout (16
  bytes on x64), providing guaranteed standard C++ inheritance layout
  for downcasting.
* For immortal static exception objects (std::bad_alloc,
  std::bad_exception), __exception_ptr_static::__on_zero_shared() is
  overridden as a no-op.
* For dynamically captured heap exceptions,
  __exception_ptr_normal::__on_zero_shared() invokes the SEH object
  destructors (pmfnUnwind / WinRT Release()) and frees heap memory
  (std::free(this)).

petrhosek committed:
Merge remote-tracking branch 'origin/main' into microsoft-excptptr

petrhosek committed:
Fix the test failures by readding the swap forward declaration

petrhosek committed:
Fix reference count leak in std::rethrow_exception

ldionne commented on libcxx/include/__exception/exception_ptr.h:
```diff
@@ -62,8 +62,6 @@ _LIBCPP_OVERRIDABLE_FUNC_VIS __cxa_exception* __cxa_init_primary_exception(
 _LIBCPP_BEGIN_UNVERSIONED_NAMESPACE_STD
 _LIBCPP_BEGIN_EXPLICIT_ABI_ANNOTATIONS
 
-#ifndef _LIBCPP_ABI_MICROSOFT
```
I really like that this patch gives us a single implementation of the public class.

ldionne commented on libcxx/include/__exception/exception_ptr.h:
```diff
@@ -180,34 +182,7 @@ _LIBCPP_HIDE_FROM_ABI exception_ptr make_exception_ptr(_Ep) _NOEXCEPT {
 
 #else // defined(_LIBCPP_ABI_MICROSOFT)
 
-class _LIBCPP_EXPORTED_FROM_ABI exception_ptr {
```
We should release-note this change, since IIUC it's ABI-breaking on Windows (since `exception_ptr` changed layout, right?).

petrhosek replied:
Updated release notes.

ldionne reviewed (approved):
Minor comments, but LGTM. I didn't review the implementation (`exception_pointer_msvc.ipp`) in detail since that part is coming from MSVC itself.

petrhosek committed:
Replace std::abort with terminate

petrhosek committed:
Update release notes

frederick-vs-ja commented on libcxx/src/support/runtime/exception_pointer_msvc.ipp:
```diff
@@ -7,70 +7,453 @@
-  if (__ptr_ == nullptr)
+  if (__ptr_ == nullptr) {
     terminate();
+  }
```
These lines don't need to be touched. I _guess_ it's better to avoid unrelated changes.

petrhosek replied:
Done.

petrhosek committed:
Omit unnecessary changes

mstorsjo commented on libcxx/docs/ReleaseNotes/22.rst:
```diff
@@ -128,6 +128,9 @@ Potentially breaking changes
   ``_LIBCPP_ABI_NO_ITERATOR_BASES``. If you are using this flag and care about ABI stability, you should set
   ``_LIBCPP_ABI_NO_REVERSE_ITERATOR_SECOND_MEMBER`` as well.
 
+- ``std::exception_ptr`` on Windows no longer relies on STL and is now implemented entirely within libc++.
```
This release note was added in the file for release 22, while the 22 release already is done. I presume this should be moved to the file for release 23?

Secondly, I'd request to clarify the terminology here. The MSVC C++ ABI is not the only ABI we support on Windows. In mingw mode, we never relied on the MS STL.

So when moving the release note to the 23 file, can you also reword this to say "in MSVC mode" instead of "on Windows", or something along those lines?

petrhosek replied:
Thanks for catching this, I created https://github.com/llvm/llvm-project/pull/208883 to move the notes.
